river/cluster/denstream.py

```diff
@@ -0,0 +1,491 @@
+    memory.
+
+    The training of a new point `p` is performed based on the following algorithm, which
+
+    * Try to merge `p` into either the nearest `p-micro-cluster`, `o-micro-cluster`, or create a new `o-micro-cluster`
+    and insert it into the outlier buffer.
+
+    * For each `T_p` iterations, consider the weights of all `p-` and `o-` micro clusters. If their weights are smaller
+    than a certain threshold (two different thresholds for two types of micro clusters), the micro cluster will be deleted.
+
+    * When a generating cluster resquest arrives, a variant of the DBSCAN algorithm [^2] from Ester et al. will be applied.
+    Finally, the label of the new point will be determined by finding the nearest cluster with respect to that point.
+
+    Parameters
+    ----------
+    decaying_factor
+        Parameter that controls the importance of historical data to current cluster.
+
+    core_weight_threshold
+        Parameter to determine the threshold of outlier relative to core micro clusters.
+
+    tolerance_factor
+        Parameter to determine the threshold of outliers relative to core micro cluster. In a normal setting,
+        this parameter is usuallly set within the range $[0,1]$.
+
+    radius
+        This parameter is passed onto the DBSCAN offline algorithm as the $\\epsilon$ parameter
+        when a clustering request arrives.
+
+    Attributes
+    ----------
+    n_clusters
+        Number of clusters generated by the algorithm.
+
+    clusters
+        A set of final clusters of type `MicroCluster`, which means that these cluster include all the required
+        information, including number of points, creation time, weight, (weighted) linear sum, (weighted) square sum,
+        center and radius.
+
+    centers
+        Centroids (central positions) of each cluster.
+
+    p_micro_clusters
+        The p micro clusters that are generated by the algorithm. When a generating cluster request arrives,
+        these p-micro-clusters will go through a variant of DBSCAN algorithm to determine the final clusters.
+
+    o_micro_clusters
+        The outlier buffer, separating the processing of the potential core-micro-cluster and outlier-micro-clusters.
+
+    References
+    ----------
+    [^1]: Feng et al (2006, pp 328-339). Density-Based Clustering over an Evolving Data Stream with Noise.
+          In Proceedings of the Sixth SIAM International Conference on Data Mining,
+          April 20–22, 2006, Bethesda, MD, USA.
+    [^2]: Ester et al (1996). A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases
+          with Noise. In KDD-96 Proceedings, AAAI.
+
+    Examples
+    ----------
+
+    The following example uses the default parameters of the algorithm to test its functionality. It can easily be seen
+    that the set of evolving points `X` are designed so that there can be a clear picture drawn on how the clusters can
+    be generated.
```

Comment:
You can easily see that the set of points can form two different clusters with the centers at `[1,1]` and `[4,2]`. However, since there are too few points, the algorithm can not generate two respective clusters.

Reply:
Got it, thanks.